### ISSUE TITLE
Forfity approves operation with Invalid Password if key is stored in fortifyapp tool cache

### ISSUE THREAD

**Issue by pmhatre-swi** [PeculiarVentures/fortify#543]
We are approving operations with Fortify Key. 
Operation is approved with Valid PIN / Password for first instance. Key details are stored at tools.fortifyapp.com after first valid approval.
Next instances onward, All operations are approved even though Invalid PIN / Password is entered. 

Details are provided below:

OS -Windows 10
Browser - Firefox 111.0.1
Yubikey - YubiKey 5C
![Invalid password](https://user-images.githubusercontent.com/129933066/230004695-b7946584-3609-4be5-acf0-537bd8ba0eb5.jpg)

![tools fortifyapp cache](https://user-images.githubusercontent.com/129933066/230004918-81c1692a-768b-4f9d-b95a-be116948802e.jpg)

**Comment by Jandrieu31** [comment]
Hi, Could we have some help regarding this issue please ? @rmhrisk @microshine Thanks

**Comment by rmhrisk** [comment]
We will need logs.

<img width="1467" alt="image" src="https://github.com/PeculiarVentures/fortify/assets/1619279/84619cf8-8ddd-49e5-84f3-c9c2323d103c">

Fortify caches the handle provided by the Yubikey middleware, not passwords.

It is strange that it appears that the Yubikey is being loaded through Windows and the Windows GIDS middleware. The log should explain why. 

Fortify prefers PKCS#11 libraries and falls back to CAPI when it can. It uses a [json file](https://github.com/PeculiarVentures/fortify/releases/download/1.8.4/card.jws) that consists of a list of smart cards and what middleware to use with them to decide what to do. Maybe the PKCS#11 does not exist or the Yubikey is provisioned differently.

Ryan

**Comment by pmhatre-swi** [comment]
Please find the attached logs.
[Fortify Logs - Forfity approves operation with Invalid Password if key is stored in fortifyapp tool cache.txt](https://github.com/PeculiarVentures/fortify/files/12181683/Fortify.Logs.-.Forfity.approves.operation.with.Invalid.Password.if.key.is.stored.in.fortifyapp.tool.cache.txt)

Here, Fortify approved the operation even after entering invalid key.

**Comment by Jandrieu31** [comment]
@rmhrisk following our discussion last month, would you mind looking into the logs you asked us to provide ? Many thanks

**Comment by microshine** [comment]
Sorry for my late response. I've reviewed the information and logs you provided and noticed the following points.

Your screenshot shows the site tools.fortifyapp.com, but it's using 'Windows CryptoAPI' as a provider. This provider relies on accessing a user's store of certificates and keys through the Windows API. It's likely that your token registers the key and certificate in the Windows store, which is why you can see it among the Windows certificates. To check this, you can:
- Insert your key into a reader device
- Open the user's certificate console (instructions available at [Microsoft's documentation](https://learn.microsoft.com/en-us/dotnet/framework/wcf/feature-details/how-to-view-certificates-with-the-mmc-snap-in#view-certificates-in-the-mmc-snap-in))

In the log file you provided, I noticed the following:
- You're trying to enter a PIN for 'GIDS card (UserPIN)'
- You're doing this not on tools.fortifyapp.com, but on the domain qa.airvantage.io.
- After entering the PIN incorrectly and getting a 'CKR_PIN_INCORRECT' error, you call the API to get the provider's certificates and keys.

Keep in mind, Fortify can retrieve keys and certificates from the token without entering a PIN, but in this case, it will only return public objects. Usually, these are certificates and public keys. To access and use private keys, you'll need to enter the correct PIN.

Additionally, you can try to access the 'GIDS card (UserPIN)' provider through tools.fortifyapp.com. If the correct PIN is entered, it will display a list of certificates; otherwise, it won’t. It's important to note that tools.fortifyapp.com only displays certificates associated with private keys. If it's not possible to retrieve the private key for a certificate, that certificate will not be shown.

**Comment by rmhrisk** [comment]
To help debug issues like this in the future we have updated the user interface to show which cryptographic provider is in use. 

<img width="797" alt="image" src="https://github.com/PeculiarVentures/fortify/assets/1619279/182353b3-483a-40ae-82ee-0d99e55ea4c3">

As you can see under the subject we now show the provider name associated with the certificate. It is our hope that this will make it clearer why a given certificate is being shown so that issues can be more easily identified in the future.

**Comment by pmhatre-swi** [comment]
Hi @microshine / @rmhrisk ,

I see that Certificates and Keys are stored in the Windows Store.
![image](https://github.com/PeculiarVentures/fortify/assets/129933066/ded91cec-9e4b-440e-94e0-1a278b33948d)

Also I see that Fortify is accessed by below sites.
![image](https://github.com/PeculiarVentures/fortify/assets/129933066/2bed6885-3c57-4380-9d60-56fdac17e807)

Below are the use-case details.
1. Root Certificate is installed on the device and Client certificate is installed on Yubikey Smartcard [https://www.yubico.com/products/yubikey-5-overview/#]
2. Devices are accessed via https://eu.airvantage.net/, similarly Operation approval with Fortify happens through the same site.
3. Client certificates are stored on Yubikey on 2 different slots [Slot 9A- Required PIN only once / Slot 9C: Requires PIN for every approval]
4. If Operation is approved using Slot 9C certificate, PIN is required for every approval, even if the Key is stored on tools.fortifyapp.com under Windows Crypto API.
5. If Operation is approved using Slot 9A certificate, PIN is required only for first time, the key gets stored on tools.fortifyapp.com under Windows Crypto API.
6. We are currently using Windows devices for our testing.

Any operation approved using Slot 9A Client Certificate no longer need any PIN and even approves operation with invalid PIN.
We want to check where exactly the certificate gets stored and how fortify plays a role in approving the operation with empty key or invalid key.

@microshine: Can you please provide the steps to access the 'GIDS card (UserPIN)' provider through tools.fortifyapp.com.
@rmhrisk : Can you please provide the steps to filter out windows stored keys and check what are the things directly returned by the token?

**Comment by microshine** [comment]
Upon further analysis of the Fortify application, I've managed to replicate a similar issue. I have opened a separate ticket at https://github.com/PeculiarVentures/webcrypto-local/issues/281 that details the identified problem.

To prevent this behavior with the PIN prompt, you need to implement an additional check on your client side:

```javascript
const loggedIn = await provider.isLoggedIn();
if (!loggedIn) {
  await provider.login();
}
```

**Comment by RafaelAban** [comment]
Hello @microshine, thanks for hint when i call to

` await event.detail.socketProvider.getCrypto(event.detail.providerId);`

the PIN prompt is launched automatically.

do you know if exist a function to log out similar to ` await provider.login(); `?

thanks.

**Comment by RafaelAban** [comment]
Hello I was doing some investigation about the issue and looks like after the first loggin the session persist logged till the yubikey is unplugged and pluged again.

PIN prompt is asking for the password each time i run 
` await event.detail.socketProvider.getCrypto(event.detail.providerId);`
but since session is logged in is not validating the password.

Imnot sure is there a way to log out or a way that
 ` await event.detail.socketProvider.getCrypto(event.detail.providerId);`
don´t ask for the PIN.

Thanks !!!

**Comment by microshine** [comment]
@RafaelAban After using the PKCS#11 session, it's recommended to end it, which helps secure the use of this session across different domains. In addition to the `login` method, the crypto object has a `logout` method that closes the session (https://peculiarventures.github.io/webcrypto-local/classes/client_src.SocketCrypto.html#logout).

Regarding the invocation of the `getCrypto` method with the Log In popup, I've reviewed the source code and don't see this method triggering the popup.

Maybe there's a chance for a call to demonstrate under what conditions this occurs? It would give me a better understanding of the issue. Here's my email microshine@peculiarventures.com we can arrange a time for the call.

**Comment by RafaelAban** [comment]
Hello @microshine let me review the links that you sent me.
 I'll send you an email to address a meeting, thank you so much.
